### DISCUSSION TITLE
CRUDApi return values

### DISCUSSION THREAD

**Discussion by obermeier** [apache/streampipes#1468] [Ideas]
Some REST calls return messages. E.g. `POST /api/v2/pipelines` returns a [SuccessMessage](https://github.com/apache/streampipes/blob/c33607da7e5b86d643db0b13941f538186711c3d/streampipes-rest/src/main/java/org/apache/streampipes/rest/impl/PipelineResource.java#L175) document.

This messages are currently not accessible by the Java-Api user. Since the interface does not return this information.

What do you think about the idea to extend the [org.apache.streampipes.client.api.CRUDApi<K, V>]( https://github.com/apache/streampipes/blob/c33607da7e5b86d643db0b13941f538186711c3d/streampipes-client/src/main/java/org/apache/streampipes/client/api/CRUDApi.java#L28)to provide this Message to the developer?

  ```void create(V element);``` vs. ```R create(V element);```  
  
This influences the other implementations of this Interface. 
Does this Message make sense for `DataLakeMeasurementApi`, `DataProcessorApi`, `DataSinkApi`, `DataStreamApi `...

**Comment by dominikriemer** [comment]
Good question - I think this makes sense and we could check that for the other classes as well. In addition, we should also think about returning `ErrorMessage` objects when something goes wrong in addition to the HTTP status.
Do the other APIs also return success messages or empty responses with an 200 status?

**Comment by obermeier** [comment]
@dominikriemer  very good point

In the current implementation delete has a [Message.class](https://github.com/apache/streampipes/blob/dev/streampipes-model/src/main/java/org/apache/streampipes/model/message/Message.java) as response with subtypes  SuccessMessage and ErrorMessage . 

If delete is used jackson could not create the message becuase it is abstract.
```Cannot construct instance of `org.apache.streampipes.model.message.Message` (no Creators, like default constructor, exist)```

Is there a mechanism in jackson to instanciate the matching class?

Do you prefer to return the generic Message or a subtype (Adding some logic to find  which type fits best)

The signature would be something like this.
```org.apache.streampipes.model.message.Message create(V element);```
